### ISSUE TITLE
Output

### ISSUE THREAD

**Issue by faren-f** [DeepGraphLearning/ULTRA#5]
For practical reasons, we need to have the predicted edges for the test data. But it seems we only access the evaluation metrics. Could you please give us guidance on how we can read the predicted edges?

**Comment by migalkin** [comment]
Here in the evaluation function we get the logits of tail predictions `t_pred`:

https://github.com/DeepGraphLearning/ULTRA/blob/f159e556fc4a77f20e9a3823f8a251eabbc075b6/script/run.py#L133-L135

Its output shape is `[batch_size, num_nodes]` and, for each `(h,r,?)` query in the batch, contains a scalar score for each entity in the graph.

You might want then to filter out scores of nodes whose edges already exist in the graph - we do that by masking those scores:

https://github.com/DeepGraphLearning/ULTRA/blob/f159e556fc4a77f20e9a3823f8a251eabbc075b6/script/run.py#L138-L141

**Comment by faren-f** [comment]
Thanks. But now the question is, how we can find the correspondence between the columns of ```t_pred``` / ```h_pred``` and the nodes in the graph.

**Comment by migalkin** [comment]
Did you check the code before opening the issue? :) 

The nodes and relations are indexed upon data loading [here](https://github.com/DeepGraphLearning/ULTRA/blob/f159e556fc4a77f20e9a3823f8a251eabbc075b6/ultra/datasets.py#L258) in respective dictionaries.

We do not save the actual entity2id and relation2id dicts into the processed folders for the sake of saving space after processing 50+ graphs, but you can easily modify the procedure to retain and load the dicts as well.
